### PR TITLE
feat/multiple output formats

### PR DESCRIPTION
**Pull request by salzig**
# Tasks
- [x] Extraction of Console output into Presenter Classes
- [x] Initial junit presenter for easier gitlab/jenkins integration. 
- [ ] Documentation on how to inject custom Presenter (if needed)

# Usage
```
bundle-audit check --format=default
bundle-audit check --format=junit
```

# Examples 
## JUnit XML 
```
<?xml version="1.0" encoding="UTF-8" ?>
<testsuites id="1547826779" name="Bundle Audit" tests="225" failures="1262">
  <testsuite id="Gemfile" name="Ruby Gemfile" failures="4">
    <testcase id="activejob" name="UNKNOWN activejob(5.2.1) Broken Access Control vulnerability in Active Job">
      <failure message="Broken Access Control vulnerability in Active Job" type="">
Name: activejob
Version: 5.2.1
Advisory: CVE-2018-16476
Criticality: Unknown
URL: https://groups.google.com/forum/#!topic/rubyonrails-security/FL4dSdzr2zw
Title: Broken Access Control vulnerability in Active Job
Solution: upgrade to ~> 4.2.11, ~> 5.0.7.1, ~> 5.1.6.1, >= 5.2.1.1
      </failure>
    </testcase>
    <testcase id="activestorage" name="UNKNOWN activestorage(5.2.1) Bypass vulnerability in Active Storage">
      <failure message="Bypass vulnerability in Active Storage" type="">
Name: activestorage
Version: 5.2.1
Advisory: CVE-2018-16477
Criticality: Unknown
URL: https://groups.google.com/forum/#!topic/rubyonrails-security/3KQRnXDIuLg
Title: Bypass vulnerability in Active Storage
Solution: upgrade to >= 5.2.1.1
      </failure>
    </testcase>
    <testcase id="rack" name="UNKNOWN rack(2.0.5) Possible DoS vulnerability in Rack">
      <failure message="Possible DoS vulnerability in Rack" type="">
Name: rack
Version: 2.0.5
Advisory: CVE-2018-16470
Criticality: Unknown
URL: https://groups.google.com/forum/#!topic/ruby-security-ann/Dz4sRl-ktKk
Title: Possible DoS vulnerability in Rack
Solution: upgrade to >= 2.0.6
      </failure>
    </testcase>
    <testcase id="rack" name="UNKNOWN rack(2.0.5) Possible XSS vulnerability in Rack">
      <failure message="Possible XSS vulnerability in Rack" type="">
Name: rack
Version: 2.0.5
Advisory: CVE-2018-16471
Criticality: Unknown
URL: https://groups.google.com/forum/#!topic/ruby-security-ann/NAalCee8n6o
Title: Possible XSS vulnerability in Rack
Solution: upgrade to ~> 1.6.11, >= 2.0.6
      </failure>
    </testcase>
  </testsuite>
</testsuites>
```

## Gitlab JUnit integration
bildschirmfoto 2019-01-18 um 16 54 56

related issues: #161

released as fork under:  https://rubygems.org/gems/bundler-budit

### CODE REVIEW

**Comment by salzig** [comment]
@postmodern i'm unsure about the cli flat, should i refactor it into `--format` before merge? Any opinion?

**Comment by postmodern** [comment]
I would rewrite this as a [Format class](https://github.com/rubysec/bundler-audit/blob/json-format/lib/bundler/audit/cli/formats/json.rb) located in `lib/bundler/audit/cli/formats/` as a separate gem, which the new bundler-audit Formats API can auto-load.

**Comment by salzig** [comment]
@postmodern just go ahead and merge your changes and i'll do another pull request. But please release it sooner then later :)

**Comment by postmodern** [comment]
@salzig I will try to allocate more time for OSS. I probably will not merge the JUnit format unless there's popular demand for it. Seems a bit out of scope, plus I designed the format API to allow loading 3rd party gems that contain a `lib/bundler/audit/cli/formats/#{format}.rb` file.

**Comment by salzig** [comment]
Is there any progress?

**Comment by Piioo** [comment]
@postmodern How far are you with your changes, this feature is interesting for me too

**Comment by postmodern** [comment]
Still busy at [commercial] work, so the status remains the same, still needs more specs.

**Comment by salzig** [comment]
> still needs more specs.

that's something you could have told me earlier. Anything specific you're missing? Anything more you didn't tell me?

**Comment by postmodern** [comment]
There are like four separate PRs for JSON/XML/HTML output. I swear I've mentioned this before in one of them, but I created [json-format](https://github.com/rubysec/bundler-audit/tree/json-format) branch which also contains the new [CLI::Formats](https://github.com/rubysec/bundler-audit/blob/json-format/lib/bundler/audit/cli/formats.rb) API and new [Results](https://github.com/rubysec/bundler-audit/tree/json-format/lib/bundler/audit/results) classes. The [Results::UnpatchedGem](https://github.com/rubysec/bundler-audit/blob/json-format/lib/bundler/audit/results/unpatched_gem.rb) needs specs, as well as fixtures for the `Gem::Specification` and `Advisory` objects. Also, [CLI::Formats::Text](https://github.com/rubysec/bundler-audit/blob/json-format/lib/bundler/audit/cli/formats/text.rb) and [CLI::Formats::JSON](https://github.com/rubysec/bundler-audit/blob/json-format/lib/bundler/audit/cli/formats/json.rb) probably need integration specs; which should be easy since we pass in `STDOUT` as an optional argument. Then, I can merge the branch, release, and others can create their own `bundler-audit-#{format}` gems for additional formats.

**Comment by hadees** [comment]
@postmodern the `Junit` one is a good standard that a lot of other tools support.  Then you can do  [junit-merge](https://www.npmjs.com/package/junit-merge) and combine them.  I'd love to see it included.  I was just looking for that ability myself so I can have the output work on CircleCI.

**Comment by postmodern** [comment]
For those that are interested, I finally got some time to sit down and write some more specs for [Results::UnpatchedGem](https://github.com/rubysec/bundler-audit/blob/json-format/spec/results/unpatched_gem_spec.rb) and added some place-holder specs for [Formats::Text](https://github.com/rubysec/bundler-audit/blob/json-format/spec/cli/formats/json_spec.rb) and [Formats::JSON](https://github.com/rubysec/bundler-audit/blob/json-format/spec/cli/formats/text_spec.rb). Almost ready to merge into the 0.7.0 into the 1.0.0 branch.

**Comment by salzig** [comment]
It's more then a year later... in the meantime we released https://rubygems.org/gems/bundler-budit and have been using that.

**Comment by postmodern** [comment]
@salzig I'm doing this in my free time outside of work. While it's good that you got a solution that works for you, I wanted to implement an extensible formatting API so people could add support for their output formats.

**Comment by salzig** [comment]
@postmodern it's not about the time you like or can spend into this project, but the fact that you prefer to tell me that my invested time is worthless. As someone who tried to contribute it feels like an contribution can't be good enough to be merged, as it's not written by yourself. 

Btw, this merge request also allowed to add custom formatters, cause the passed in presenter name is fetched as constant name from the module `Presenter`. Anyone could potentially create a `Presenter::Some` and pass `--presenter=some` as parameter.

**Comment by postmodern** [comment]
@salzig while it's difficult to reject PRs, I'd rather not take on the responsibility to maintain multiple formats, which is why I had to reject other format PRs. With the Formats API custom formats can be distributed as separate gems that are required/registered. Also, you probably want to use `const_get(..., false)` to prevent it from accidentally loading global constants.

**Comment by postmodern** [comment]
Closing this. The 0.8.0 branch will soon be merged/released. Afterwards a `bundler-audit-junit` library can be created which provides a `Bundler::Audit::CLI::Formats::JUnit` format (see: [Bundler::Audit::CLI::Formats::JSON](https://github.com/rubysec/bundler-audit/blob/0.8.0/lib/bundler/audit/cli/formats/json.rb) for example). The library can also dep-in an XML building gem or use ERB+`CGI.escapeHTML`.

File: lib/bundler/audit/presenter/junit.rb [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+  <testsuite id="Gemfile" name="Ruby Gemfile" failures="<%= @advisory_bundles.size %>">
+    <%- @advisory_bundles.each do |bundle| -%>
+    <testcase id="<%= bundle.gem.name %>" name="<%= bundle.gem.name %>">
+      <failure message="<%= bundle.advisory.title %>" type="<%= bundle.advisory.criticality %>">
+Name: <%= bundle.gem.name %>
+Version: <%= bundle.gem.version %>
+Advisory: <%= advisory_ref(bundle.advisory) %>
+Criticality: <%= advisory_criticality(bundle.advisory) %>
+URL: <%= bundle.advisory.url %>
+Title: <%= bundle.advisory.title %>
```
Comment (postmodern) [inline lib/bundler/audit/presenter/junit.rb:55]: FYI `Advisory#title` can contain XML/HTML special characters. Always use something like `CGI.escapeHTML` when outputting free-form text into XML.

**Pull request closed without merging**